setup.py

```diff
@@ -24,13 +24,48 @@
 import subprocess
 import textwrap
 import warnings
+import versioneer
+import builtins
+
 
 
+# This is a bit hackish: we are setting a global variable so that the main
+# numpy __init__ can detect if it is being loaded by the setup routine, to
+# avoid attempting to load components that aren't built yet.  While ugly, it's
+# a lot more robust than what was previously being used.
+builtins.__NUMPY_SETUP__ = True
+
 if sys.version_info[:2] < (3, 7):
     raise RuntimeError("Python version >= 3.7 required.")
 
-import builtins
+# The first version not in the `Programming Language :: Python :: ...` classifiers above
+if sys.version_info >= (3, 10):
+    warnings.warn(
+        f"NumPy {VERSION} may not yet support Python "
+        f"{sys.version_info.major}.{sys.version_info.minor}.",
+        RuntimeWarning,
+    )
+
+# BEFORE importing setuptools, remove MANIFEST. Otherwise it may not be
+# properly updated when the contents of directories change (true for distutils,
+# not sure about setuptools).
+if os.path.exists('MANIFEST'):
+    os.remove('MANIFEST')
 
+# We need to import setuptools here in order for it to persist in sys.modules.
+# Its presence/absence is used in subclassing setup in numpy/distutils/core.py,
+# which may not be the most robust design.
+import setuptools
```

Comment:
This was tricky to discover, we have been doing it wrong all along.

Reply:
I think the rationale is that setuptools has to be imported before distutils; therefore, if `setuptools` is not already in `sys.modules` before numpy.distutils is imported, it's too late. I thought we already handled this, as I remember warnings being emitted about it...

Reply:
It caused a problem with sdist that was subclassed before the import. The root problem may have been due to initializing the `cmdclass` with `versioneer.get_cmdclass` that took as its argument the cmdclass from `numpy/distutils/core.py`.  We will see if any other problems show up.